README.md
# CLOUDLESS LABS - AWSX

This package exposes APIs that wrap the AWS SDK. Its purpose is to get started faster with the AWS SDK for the most common scenarios. The AWS SDK is also exposed to allow access to the native APIs.

```
npm i @cloudlessopenlabs/awsx
```

# Table of contents

> * [APIs](#apis)
>	- [Cloudfront](#cloudfront)
>		- [`cloudfront.distribution.exists`](#cloudfrontdistributionexists)
>		- [`cloudfront.distribution.select` and `cloudfront.distribution.find`](#cloudfrontdistributionselect-and-cloudfrontdistributionfind`)
>		- [`cloudfront.distribution.create`](#cloudfrontdistributioncreate)
>		- [`cloudfront.distribution.invalidate`](#cloudfrontdistributioninvalidate)
>		- [`cloudfront.distribution.update`](#cloudfrontdistributionupdate)
>	- [Cloudwatch](#cloudwatch)
>		- [`cloudwatch.logs.select`](#cloudwatchlogsselect)
>	- [Parameter Store](#parameter-store)
>		- [`parameterStore.get`](#parameterstoreget)
>		- [`parameterStore.put`](#parameterstoreput)
>	- [Resource](#resource)
>	- [S3](#s3)
>		- [`s3.bucket.exists`](#s3bucketexists)
>		- [`s3.bucket.list`](#s3bucketlist)
>		- [`s3.bucket.get`](#s3bucketget)
>		- [`s3.bucket.setWebsite`](#s3bucketsetWebsite)
>		- [`s3.object.get`](#s3objectget)
>		- [`s3.object.put`](#s3objectput)
>		- [`s3.object.upload`](#s3objectupload) 
>		- [`s3.object.sync`](#s3objectsync)
>		- [`s3.object.remove`](#s3objectremove)
>	- [SNS](#sns)
>		- [`topic.publish`](#topicpublish)
> * [Annexes](#annexes)
>	- [Cloudfront distribution with S3 static website bucket](#cloudfront-distribution-with-s3-static-website-bucket)
>	- [Cloudfront distribution with private S3 bucket ](#cloudfront-distribution-with-private-s3-bucket)
> * [License](#license)

# APIs
## Cloudfront

For concrete examples, please refer to the [Annexes](#annexes):
- [Cloudfront distribution with S3 static website bucket](#cloudfront-distribution-with-s3-static-website-bucket)
- [Cloudfront distribution with private S3 bucket ](#cloudfront-distribution-with-private-s3-bucket)

### `cloudfront.distribution.exists`

```js
const { error: { catchErrors, wrapErrors } } = require('puffy-core')
const { cloudfront } = require('@cloudlessopenlabs/awsx')

const DISTRO = `my-distro-name`

const main = () => catchErrors((async () => {
	// ID exists
	const [idExistsErrors, idExists] = await cloudfront.distribution.exists({ 
		id: '12345'
	})
	if (idExistsErrors)
		throw wrapErrors(`Failed to find by ID`, idExistsErrors)
	else 
		console.log(`ID exists`)

	// Distro with tag exists
	const [tagExistsErrors, tagExists] = await cloudfront.distribution.exists({ 
		tags: { Name:DISTRO }
	})
	if (tagExistsErrors)
		throw wrapErrors(`Failed to find by ID`, tagExistsErrors)
	else 
		console.log(`Tag exists`)
```

### `cloudfront.distribution.select` and `cloudfront.distribution.find`

`find` and `select` uses the same API. `find` returns an object while `select` returns an array.

```js
const { error: { catchErrors, wrapErrors } } = require('puffy-core')
const { cloudfront } = require('@cloudlessopenlabs/awsx')

const DISTRO = `my-distro-name`

const main = () => catchErrors((async () => {
	// Find by ID
	const [distroErrors, distro] = await cloudfront.distribution.find({ 
		id: '12345'
	})
	if (distroErrors)
		throw wrapErrors(`Failed to find by ID`, distroErrors)
	else if (distro)
		console.log(distro)
		// {
		// 	id: 'E2WJO325O501XD',
		// 	arn: 'arn:aws:cloudfront::084126072180:distribution/E2WJO325O501XD',
		// 	domainName: 'dyoumeptyjf92.cloudfront.net',
		// 	status: 'InProgress',
		// 	lastUpdate: 2021-10-15T08:40:11.908Z,
		// 	eTag: 'E1AQWRQF1J4O48',
		// 	origin: {
		// 		domain: 'nic-today-20211015.s3.ap-southeast-2.amazonaws.com',
		// 		type: 's3'
		// 	},
		// 	aliases: [],
		// 	enabled: true
		// }
	else
		console.log(`Distro not found`)

	// Find by tag
	const [distro2Errors, distro2] = await cloudfront.distribution.find({ 
		tags: { Name:DISTRO }
	})
	if (distro2Errors)
		throw wrapErrors(`Failed to find by tag`, distro2Errors)
	else if (distro2)
		console.log(distro2)
		// {
		// 	id: 'E2WJO325O501XD',
		// 	arn: 'arn:aws:cloudfront::084126072180:distribution/E2WJO325O501XD',
		// 	domainName: 'dyoumeptyjf92.cloudfront.net',
		// 	status: 'InProgress',
		// 	lastUpdate: 2021-10-15T08:40:11.908Z,
		// 	eTag: 'E1AQWRQF1J4O48',
		// 	origin: {
		// 		domain: 'nic-today-20211015.s3.ap-southeast-2.amazonaws.com',
		// 		type: 's3'
		// 	},
		// 	aliases: [],
		// 	enabled: true
		// }
	else
		console.log(`Distro not found`)
```

### `cloudfront.distribution.create`

```js
const { error: { catchErrors, wrapErrors } } = require('puffy-core')
const { cloudfront } = require('@cloudlessopenlabs/awsx')

const DISTRO = `my-distro-name`

const main = () => catchErrors((async () => {
	const [distroErrors, distro] = await cloudfront.distribution.create({
		name: DISTRO,
		domain: 'my-bucket-website.s3.ap-southeast-2.amazonaws.com',
		operationId: '123456', 
		enabled: true,
		tags: { 
			Project:'Demo', 
			Env:'Dev',
			Name: DISTRO
		}
	})

	if (distroErrors)
		throw wrapErrors(`Distro creation failed`, distroErrors)

	console.log(distro)
	// {
	// 	id: 'E2WJO325O501XD',
	// 	arn: 'arn:aws:cloudfront::084126072180:distribution/E2WJO325O501XD',
	// 	status: 'InProgress'.
	// 	domain: 'dyoumeptyjf92.cloudfront.net'
	// }
```

### `cloudfront.distribution.invalidate`

```js
const { error: { catchErrors, wrapErrors } } = require('puffy-core')
const { cloudfront } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	// Invalidate all paths
	const [invalidationErrors, invalidation] = await cloudfront.distribution.invalidate({ 
		id: 'E2WJO325O501XD', 
		operationId: `${Date.now()}`, 
		paths: ['/*']
	})
	if (invalidationErrors)
		throw wrapErrors(`Failed to invalidate distro`, invalidationErrors) 

	console.log('Distro invalidation started')
	console.log(invalidation)
	// {
	// 	location: 'https://cloudfront.amazonaws.com/2019-03-26/distribution/E3ICGTU0Z3IYAZ/invalidation/IGLGGAGD9PT2X',
	// 	invalidation: {
	// 		id: 'IGLGGAGD9PT2X',
	// 		status: 'InProgress',
	// 		createTime: 2021-10-16T04:14:34.514Z,
	// 		paths: [ '/*' ],
	// 		operationId: '1634357673513'
	// 	}
	// }
```

### `cloudfront.distribution.update`

```js
const { error: { catchErrors, wrapErrors } } = require('puffy-core')
const { cloudfront } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	// Invalidate all paths
	const [updateErrors, results] = await cloudfront.distribution.update({ 
		id:'E2WJO325O501XD', 
		// tags: { Name:'my-distro' } // find by tags
		config: { // Update
			domain:bucketWeb01.bucketRegionalDomainName
		} 
	})
	if (updateErrors)
		throw wrapErrors('Failed to update distro', updateErrors) 

	console.log('Distro update successfull')
	console.log(results)
	// {
	// 	updateOccured: true, // False means that the last config is the same is the new, meaning there was no need for an update.
	// 	config: {
	// 		... 
	// 	}
	// }
```

## Cloudwatch
### `cloudwatch.logs.select`

```js
const { error: { catchErrors, wrapErrors } } = require('puffy-core')
const { cloudwatch } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	// Invalidate all paths
	const [updateErrors, results] = await cloudwatch.logs.select({ 
		where: {
			logGroup: '/aws/lambda/some-log-group',
			timeRange: {
				last: {
					value: 3,
					unit: 'hour'
				},
				// startDateUTC: '2022-01-01',
				// endDateUTC: '2022-01-04'
			}
		},
		// limit: 1000, // default 10,000
		all: true // Default false. True means that it keeps querying data until it gets it all (one request returns max 10K records or 1MB worth of data`)
	})
	if (updateErrors)
		throw wrapErrors('Failed to get logs', updateErrors) 

	console.log('Logs acquired successfully')
	console.log(results)
	// {
 	//		count: 120,
 	//		events: [{
 	//			logStreamName: 'dwdw',
 	//			timestamp: 2022-01-29T07:38:10.053Z,		// UTC time when the log was recorded			
 	//			message: 'ewdew',
 	//			ingestionTime: 2022-01-29T07:38:10.053Z,	// UTC time when the log added to CloudWatch
 	//			eventId: '123544454'
 	//		},
 	//		iteration: 1,									//Nb. of API requests to get all the data. (max. 100).
 	//		message: null,									// Warning message if max iteration reached.
 	//		nextToken: null
 	// }
```

## Parameter Store
### `parameterStore.get`

```js
const { parameterStore } = require('@cloudlessopenlabs/awsx')

parameterStore.get({
	name: 'my-parameter-store-name',
	version: 2, // Optional. If not defined, the latest version is used.
	json: true // Optional. Default false.
}).then(([errors, resp]) => console.log(resp))
// {
// 	name: 'my-parameter-store-name',
// 	type: 'String',
// 	value: {
// 		hello: 'World'
// 	},
// 	version: 2,
// 	lastModifiedDate: 2022-01-30T07:25:07.516Z,
// 	arn: 'arn:....',
// 	dataType: 'text'
// }
```

To use this API, the following policy must be attached to the hosting environmnet's IAM role:

```js
{
	Version: '2012-10-17',
	Statement: [{
		Action: [
			'ssm:GetParameter'
		],
		Resource: '*',
		Effect: 'Allow'
	}]
}
```

### `parameterStore.put`

```js
const { parameterStore } = require('@cloudlessopenlabs/awsx')

parameterStore.put({
	name: 'my-parameter-store-name'
	value: {
		hello: 'World'
	},
	// description: 'dewde'
	overWrite: 'dewde'
	// tier: 'Advanced', // 'Standard' (default) | 'Advanced' | 'Intelligent-Tiering'
	tags: {
		Name: 'hello'
	}
}).then(([errors, resp]) => console.log(resp))
// {
// 	version: 1,
// 	tier: 'Standard'
// }
```

## Resource

> __WARNING__: Certain AWS services are global (e.g., 'cloudfront'), which means their region is 'us-east-1'. 		

```js
const { error: { catchErrors, mergeErrors } } = require('puffy-core')
const { resource } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [resourceErrors, resources] = await resource.getByTags({ 
		tags: { // required
			Name:'my-resource-name' 
		}, 
		region: 'us-east-1', // required
		types: ['cloudfront:distribution'] // optional
	})
	if (resourceErrors)
		throw wrapErrors(`Failed to get resource by tag`, resourceErrors)

	console.log(resources)
	// {
	// 	paginationToken: '',
	// 	resources:[{
	// 		arn: 'arn:aws:cloudfront::1234567:distribution/SHWHSW3213',
	// 		tags: {
	// 			Name:'my-resource-name'
	// 		}
	// 	}]
	// }
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

## S3

For concrete examples, please refer to the [Annexes](#annexes):
- [Cloudfront distribution with S3 static website bucket](#cloudfront-distribution-with-s3-static-website-bucket)
- [Cloudfront distribution with private S3 bucket ](#cloudfront-distribution-with-private-s3-bucket)

### `s3.bucket.exists`

```js
const { error: { catchErrors, mergeErrors } } = require('puffy-core')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const bucketName = 'some-bucket-name'
	const bucketExists = await s3.bucket.exists(bucketName)
	console.log(`Bucket '${bucketName} ${bucketExists ? ' ' : 'does not '}exist'`)
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.bucket.list`

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [errors, bucketList] = await s3.bucket.list()
	if (errors)
		throw wrapErrors('Failed to list buckets', errors)
	else {
		console.log(`Found ${bucketList.buckets.length} buckets.`)
		console.log(bucketList.owner)
		if (bucketList.buckets.length) {
			console.log('First bucket')
			console.log(bucketList.buckets[0].name)
			console.log(bucketList.buckets[0].creationDate)
		}
	}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.bucket.get`

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const bucketName = 'some-bucket-name'
	const [errors, bucket] = await s3.bucket.get(bucketName, { website:true })
	if (errors)
		throw wrapErrors('Failed to get bucket', errors)
	else {
		console.log(bucket)
		// {
		// 		region: 'ap-southeast-2',
		// 		location: 'http://some-bucket-name.s3.amazonaws.com/'
		// 		regionalLocation: 'http://some-bucket-name.s3.ap-southeast-2.amazonaws.com/',
		// 		bucketDomainName: 'some-bucket-name.s3.amazonaws.com'
		// 		bucketRegionalDomainName: 'some-bucket-name.s3.ap-southeast-2.amazonaws.com',
		// 		website: true, // Only with 'options.website:true'
		// 		websiteEndpoint: 'http://some-bucket-name.s3-website-ap-southeast-2.amazonaws.com' // Only with 'options.website:true'
		// }
	}
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.bucket.setWebsite`

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const bucketName = 'some-bucket-name'
	const [errors] = await s3.bucket.setWebsite({ 
		bucket: bucketName, 
		index: 'home.html', // default 'index.html'
		error: 'error.html'
	})
	if (errors)
		throw wrapErrors('Failed to get bucket', errors)
	else {
		console.log(`Bucket set as website`)
	}
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.object.get`

Requires the `s3:GetObject` permission. Ideally you also add the `s3:ListBucket` permission. Without the `s3:ListBucket` permission, missing object return a 403 Access Denied error rather than a 404 No suck key error.

> WARNING: There used to be a bug where the underlying AWS API always return 403 access denied regardless of whether the `s3:ListBucket` permission is present or not.

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { join } = require('path')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [errors, data] = await s3.object.get({
		bucket: 'my-bucket-name',
		key: 'path/to/my-folder/example.json',
		type: 'json' // valid values: 'buffer' (default), 'json', 'text'
	})

	if (errors)
		throw wrapErrors('Failed to content to bucket', errors)

	console.log(data)
	// {
	// 	acceptRanges: 'bytes',
	// 	lastModified: 2022-01-30T09:52:30.000Z,
	// 	contentLength: 22,
	// 	eTag: '"add7403b95c4164509110b1eac281ae6"',
	// 	contentType: 'application/json',
	// 	metadata: {},
	// 	body: {
	// 		hello: 'World'
	// 	}
	// }
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.object.put`

Requires the `s3:PutObject` permission.

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { join } = require('path')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [errors, data] = await s3.object.put({
		body: { // When the 'body' is not a Buffer, it is stringified. The format can be controlled via the 'format' property.
			hello: 'world'
		},
		// format: 'json-multi-line', // This option stringifies the JSON using multiple lines rather than a single one.
		bucket: 'my-bucket-name',
		key: 'path/to/my-folder/example.json'
	})

	if (errors)
		throw wrapErrors('Failed to content to bucket', errors)

	console.log(data)
	// {
	// 	etag: '"add7403b95c4164509110b1eac281ae6"'
	// }
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.object.upload`

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { join } = require('path')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [uploadErrors, filesInDir] = await s3.object.upload({ 
		bucket: 'my-super-bucket', 
		dir: join(__dirname, './app'), 
		ignore: '**/node_modules/**', 
		ignoreObjects:[{ key:'src/bundle.js', hash:'123456' }], // If 'src/bundle.js' has not changed (i.e., its hash is the same), then don't upload it
		noWarning: true
	})
	if (uploadErrors)
		throw wrapErrors('Failed to upload files to bucket', uploadErrors)
	else {
		console.log(`${filesInDir.length} files in directory`)
		console.log(filesInDir[0].key)
		console.log(filesInDir[0].hash)
	}
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.object.sync`

Does the same as 'upload' but with the ability to delete files that have been removed locally.

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { join } = require('path')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [syncErrors, synchedData] = await s3.object.sync({ 
		bucket: 'my-super-bucket', 
		dir: join(__dirname, './app'), 
		ignore: '**/node_modules/**', 
		ignoreObjects:[{ key:'src/bundle.js', hash:'123456' }], // If 'src/bundle.js' has not changed (i.e., its hash is the same), then don't upload it
		noWarning: true
	})
	if (syncErrors)
		throw wrapErrors('Failed to sync files to bucket', syncErrors)
	else {
		console.log(synchedData)
		// {
		// 	updated: true, // True means at least one file was either uploaded or deleted.
		// 	srcFiles:[{
		// 		file: '/Users/you/Documents/my-app/bundle.js',
		// 		dir: '/Users/you/Documents/my-app/',
		// 		key: 'bundle.js',
		// 		path: '/bundle.js',
		// 		hash: '32132313213123321313',
		// 		contentType: 'application/javascript',
		// 		contentLength: '12345',
		// 		content: 'dewdwedewdeewdedewdedewdedewdeew...',
		// 	}],
		// 	uploadedFiles:[{
		// 		file: '/Users/you/Documents/my-app/bundle.js',
		// 		dir: '/Users/you/Documents/my-app/',
		// 		key: 'bundle.js',
		// 		path: '/bundle.js',
		// 		hash: '32132313213123321313',
		// 		contentType: 'application/javascript',
		// 		contentLength: '12345',
		// 		content: 'dewdwedewdeewdedewdedewdedewdeew...',
		// 	}],
		// 	deletedFiles:[]
		// }
	}
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

### `s3.object.remove`

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { s3 } = require('@cloudlessopenlabs/awsx')

const main = () => catchErrors((async () => {
	const [rmErrors] = await s3.object.remove({ 
		bucket: 'my-super-bucket', 
		keys:[
			'src/bundle.js',
			'src/bundle.map.js'
		]
	})
	if (rmErrors)
		throw wrapErrors('Failed to remove files from bucket', rmErrors)
	else
		console.log(`Files removed`)
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

## SNS
### `topic.publish`


```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { sns } = require('@cloudlessopenlabs/awsx')

const topic = new sns.Topic('arn::your-topic-arn') 

const main = () => catchErrors((async () => {
	// Supports sending string message "as-is"
	const [errors01, resp01] = await topic.publish('hello world')
	// Supports sending an object, as long as the object contains the required 'body' property.
	const [errors02, resp02] = await topic.publish({ body:'Hello world', attributes: { hello:'world' } })
	// Supports sending SMSes
	const [errors03, resp03] = await topic.publish('Hello world', { 
		phone:'+61420876543',
		// subject: 'Hello from me', // Optional
		// type: 'promotional' // Optional. Valid values: 'promotional', 'transactional'
	})

	if (errors01 || errors02 || errors03)
		throw wrapErrors(errors01 || errors02 || errors03)
	else {
		console.log(resp01) // { ResponseMetadata: { RequestId: '123' }, MessageId: '456' }
		console.log(resp02) // { ResponseMetadata: { RequestId: '321' }, MessageId: '345' }
		console.log(resp03) // { ResponseMetadata: { RequestId: '875' }, MessageId: '517' }
	}
}
})())

main().then(([errors]) => {
	if (errors)
		console.error(mergeErrors(errors).stack)
	else
		console.log('All good')
})
```

# Annexes
## Cloudfront distribution with S3 static website bucket

The 2 APIs in the code below are:
- `cloudfront.distribution.exists`
- `cloudfront.distribution.create`

Notice that the only way to use `cloudfront.distribution.exists` with another predicate than the Cloudfront distribution ID is with tagging. This means that the distro MUST be tagged. 

```js
const { error: { catchErrors, wrapErrors, mergeErrors } } = require('puffy-core')
const { join } = require('path')
const { s3, cloudfront, resource } = require('@cloudlessopenlabs/awsx')

const BUCKET = 'nic-today-20211015'
const DISTRO = `${BUCKET}-distro`
const REGION = 'ap-southeast-2'

const main = () => catchErrors((async () => {
	// 1.Making sure the bucket exists
	if (!(await s3.bucket.exists(BUCKET))) {
		console.log(`Creating bucket '${BUCKET}'...`)
		const [createErrors, newBucket] = await s3.bucket.create({ 
			name: BUCKET, 
			acl: 'public-read',  // Default 'private'
			region: REGION, // default: 'us-east-1' 
			tags: { 
				Project:'Demo', 
				Env:'Dev' 
			}
		})

		if (createErrors)
			throw wrapErrors(`Bucket creation failed`, createErrors)

		console.log(`Bucket '${BUCKET}' successfully created.`)
	} else
		console.log(`Bucket '${BUCKET}' already exists.`)

	const [getErrors, bucketDetails] = await s3.bucket.get(BUCKET, { website:true })
	if (getErrors)
		throw wrapErrors(`Bucket info failed`, getErrors)

	// 2. Making sure the website is set as a website
	if (!bucketDetails.website) {
		console.log(`Setting bucket '${BUCKET}' to website`)
		const [websiteErrors] = await s3.bucket.setWebsite({ bucket:BUCKET })
		if (websiteErrors)
			throw wrapErrors(`Failed to set bucket as website`, websiteErrors)
	} else
		console.log(`Bucket '${BUCKET}' already set to website`)

	// 3. Uploading files to the bucket
	const [syncErrors] = await s3.object.sync({
		bucket: BUCKET, 
		dir: join(__dirname, './demo'), 
		noWarning: true
	})

	if (syncErrors)
		throw wrapErrors(`Synching files failed`, syncErrors)

	// 4. Adding a cloudfront distro on the bucket
	const [distroExistsErrors, distroExists] = await cloudfront.distribution.exists({ 
		tags: { Name:DISTRO }
	})
	if (distroExistsErrors)
		throw wrapErrors(`Failed to confirm whether the distro exists or not`, distroExistsErrors)

	if (!distroExists) {
		console.log(`Creating new distro tagged 'Name:${DISTRO}' for bucket '${BUCKET}'`)
		const [distroErrors, distro] = await cloudfront.distribution.create({
			name: DISTRO,
			domain: bucketDetails.bucketRegionalDomainName,
			operationId: DISTRO, 
			enabled: true,
			tags: { 
				Project:'Demo', 
				Env:'Dev',

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.1](https://github.com/cloudlesslabs/awsx/compare/v0.2.0...v0.2.1) (2022-09-19)


### Bug Fixes

* Incorrect puffy-core dependency ([1e0b98e](https://github.com/cloudlesslabs/awsx/commit/1e0b98e4969e3d5ac0f4ef1cdfe9b4a9edecdf2c))

## [0.2.0](https://github.com/cloudlesslabs/awsx/compare/v0.1.2...v0.2.0) (2022-09-19)


### Features

* Add support for SNS ([78e2faf](https://github.com/cloudlesslabs/awsx/commit/78e2faf46aa49e07db468eadf2abc08eff5a1845))

### [0.1.2](https://github.com/cloudlesslabs/awsx/compare/v0.1.1...v0.1.2) (2022-07-20)


### Bug Fixes

* vulnerability ([5cb4dce](https://github.com/cloudlesslabs/awsx/commit/5cb4dce10d899d28425ab1c04444dfea37ce6599))

### [0.1.1](https://github.com/cloudlesslabs/awsx/compare/v0.1.0...v0.1.1) (2022-05-11)

## [0.1.0](https://github.com/cloudlesslabs/awsx/compare/v0.0.5...v0.1.0) (2022-01-31)


### Features

* Add support for formatting a JSON object when stored in S3 ([c84ed98](https://github.com/cloudlesslabs/awsx/commit/c84ed98f992ebf4c5ab53cbf1693a8f87cb37764))

### [0.0.5](https://github.com/cloudlesslabs/awsx/compare/v0.0.4...v0.0.5) (2022-01-30)


### Features

* Add support for parsing an S3 object's body from buffer to JSON or text ([5d2a9d7](https://github.com/cloudlesslabs/awsx/commit/5d2a9d7bfef1fe5984f88325b49d9394577b10f4))

### [0.0.4](https://github.com/cloudlesslabs/awsx/compare/v0.0.3...v0.0.4) (2022-01-30)


### Features

* Add support for getting S3 objects ([fc9c864](https://github.com/cloudlesslabs/awsx/commit/fc9c8644611885933bf0c59b7dc05fd6c7a5a23a))

### [0.0.3](https://github.com/cloudlesslabs/awsx/compare/v0.0.2...v0.0.3) (2022-01-30)

### 0.0.2 (2022-01-30)


### Features

* Add support for cloudfront cloudwatch parameterStore resource s3 ([2254bf7](https://github.com/cloudlesslabs/awsx/commit/2254bf78731278d774045ab9651e315b93c2ca26))


src/cloudwatch.js
/**
 * Copyright (c) 2019-2022, Cloudless Consulting Pty Ltd.
 * All rights reserved.
 * 
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
*/

const { error:{ catchErrors, wrapErrors } } = require('puffy-core')
const AWS = require('aws-sdk')
const { REGION, promisify } = require('./_utils')
const logs = new AWS.CloudWatchLogs({ apiVersion: '2014-03-28', region: REGION })

const MINUTE = 60*1000
const HOUR = 60*MINUTE
const DAY = 24*HOUR
const WEEK = 7*DAY
const MAX_ITERATION = 100

// Doc: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudWatchLogs.html#filterLogEvents-property
const _logsFilterLogEvents = promisify(logs, 'filterLogEvents', true)

/** 
 * Gets logs from a log group. The logs are sorted in ascending order based on the ingestion time (this cannot be changed).
 * 
 * Doc: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudWatchLogs.html#filterLogEvents-property
 * 
 * @param  {Object} 		query
 * @param  {Object} 			.where
 * @param  {String} 				.logGroup			Required (e.g., '/aws/lambda/lineup-workloads-project-api-dev')
 * @param  {Object}		 			.timeRange
 * @param  {Object}		 				.last			Default 3 hours
 * @param  {Number}		 					.value		Default 3
 * @param  {Number}		 					.unit		Allowed values: 'minute', 'hour' (default), 'day', 'week'
 * @param  {Date|String}		 		.startDateUTC
 * @param  {Date|String}		 	.	.endDateUTC
 * @param  {String} 				.nextToken
 * @param  {String} 				.filterPattern		(1)
 * @param  {Number} 			.limit					Default 10,000 (which is the max allowed)
 * @param  {String}		 		.sort					Allowed values: 'asc' (default), 'desc'
 * @param  {Boolean}		 	.all					Default false. True means that all values for that time range are retrieved.
 * 
 * @return {Object}			response
 * @return {Number}				.count					Total events.
 * @return {[Object]}			.events[]
 * @return {String}					.logStreamName
 * @return {Date}					.timestamp			UTC time when the log was recorded			
 * @return {String}					.message
 * @return {Date}					.ingestionTime		UTC time when the log added to CloudWatch
 * @return {String}					.eventId
 * @return {Number}				.iteration				Number of API requests performed in order to get all the data. (max. 100).
 * @return {String}				.message				Warning message if max iteration reached.
 * @return {String}				.nextToken
 *
 * (1) Example:
 * 	- 'ERROR' returns all message that contains the string 'ERROR'. Both 'ERROR 123 oh no' and 'ERRORCODE -1' would match.
 * 	- '"ERROR" "no"' returns all message that contains the string both 'ERROR' and 'no'. With Both 'ERROR 123 oh no' and 'ERRORCODE -1' only the first one would match.
 *
 * Full doc at https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html
 * 
 */
const select = query => catchErrors((async () => {
	const e = (...args) => wrapErrors('Failed to select CloudWatch logs', ...args)

	const { where, limit, sort, all } = query || {}

	let { logGroup, nextToken, filterPattern, timeRange } = where || {}
	let { last, startDateUTC, endDateUTC } = timeRange || {}

	if (!logGroup)
		throw e('Missing required argument \'query.where.logGroup\'')
	
	const params = {
		logGroupName: logGroup,
		limit: all ? 10000 : limit || 10000
	}

	if (nextToken) 
		params.nextToken = nextToken

	if (filterPattern) 
		params.filterPattern = filterPattern

	if (last) {
		const { value, unit  } = last || {}
		const t = isNaN(value*1) || value*1 <= 0 ? 3 : value*1
		const milliseconds = unit == 'minute' ? MINUTE : (!unit || unit == 'hour') ? HOUR : unit == 'day' ? DAY : unit == 'week' ? WEEK : null
		if (t && !milliseconds)
			throw e(`Wrong argument exception. 'unit' value '${unit}' is not supported.`)
		
		params.startTime = Date.now() - t*milliseconds
		params.endTime = Date.now()
	} else {
		if (!startDateUTC && !endDateUTC) {
			params.startTime = Date.now() - 3*HOUR
			params.endTime = Date.now()
		} else {
			if (startDateUTC) {
				if (!(startDateUTC instanceof Date)) {
					startDateUTC = new Date(startDateUTC)
					if (isNaN(startDateUTC))
						throw e(`Wrong argument exception. 'startDateUTC' is expected to be a date or a parseable date (e.g., '2021-12-26'). Failed to parse ${startDateUTC} to a UTC date.`)
				}
				params.startTime = startDateUTC.getTime()
			}
			if (endDateUTC) {
				if (!(endDateUTC instanceof Date)) {
					endDateUTC = new Date(endDateUTC)
					if (isNaN(endDateUTC))
						throw e(`Wrong argument exception. 'endDateUTC' is expected to be a date or a parseable date (e.g., '2021-12-26'). Failed to parse ${endDateUTC} to a UTC date.`)
				}
				params.endTime = endDateUTC.getTime()
			}
		}
	}

	// Maximum amount of data retrieved is 100 request (i.e., 100MB).
	const count = all ? MAX_ITERATION : 1
	const events = []
	let next, iteration = 0
	for (let i=0;i<count;i++) {
		iteration++
		const p = { ...params }
		if (next)
			p.nextToken = next

		const [errors, result] = await _logsFilterLogEvents(p)
		if (errors)
			throw e(`Failed to extract logs after the ${iteration}th iteration`, errors)

		next = (result||{}).nextToken
		events.push(...((result||{}).events||[]))

		if (!next)
			break
	}

	const data = {
		count: events.length,
		nextToken:next||null,
		iteration,
		message: iteration == MAX_ITERATION && next ? `WARNING: Max iteration(${MAX_ITERATION}) exceeded. Failed to extract all logs. Reduce the time range in order to fit all the data.` : null
	}

	const dir = sort == 'desc' ? -1 : 1 
	if (data.count)
		data.events = events.sort((a,b) => a.timestamp < b.timestamp ? -dir : dir).map(e => {
			const timestamp = new Date(e.timestamp)
			const ingestionTime = new Date(e.ingestionTime)
			return {
				...e,
				timestamp,
				ingestionTime
			}
		})

	return data
})())


module.exports = {
	logs: {
		select
	}
}





